### DISCUSSION TITLE
Next app bundle too large, probably three shaking?

### DISCUSSION THREAD

**Discussion by DanielVip3** [vercel/next.js#13763] [Help]
I have built a Next + Express app which goes painfully slow. Next's bundles, on build, look so big and the first load dimension is written in red.

[![Next.js build resume][1]][1]

(sorry for that weird characters, that's a problem of my SSH CLI)

Anyway, I don't know where to put my head anymore. I minify all assets with Webpack, have everything in cache on server-side and server speed and CPU doesn't seem to be a problem (that's not a very popular web app, so we don't have too much overhead).

I thought the problem is components-side, so I used Next's webpack bundle analyzer and those 2 analyzed client and server bundle files resulted, you can download them from this link:
https://gofile.io/d/HSSIRP

I mean, why the hell should the client bundle be 15 MB parsed? I am using tree shaking everywhere; my app only uses like 3 or 4 external modules, apart from Next's internal modules; the main one I use is Material-UI which is singularly imported everytime like this:

```js
import Snackbar from '@material-ui/core/Snackbar';
```
and I don't use more than like, 30 or 35 different components.

Then, I use this kind of tree shaking for all the others libraries:
```js
import { Tooltip } from 'react-tippy';
import { withRouter } from 'next/router';

/* and for ones which don't need tree shaking */
import fetch from 'isomorphic-unfetch';
```

But it seems that tree shaking isn't working, or at least it does but works in a wrong way.

I can't fully understand these webpack bundle analyzer pages, so can anyone tell me where the problem could be? It seems that most of the overhead comes from Material UI, which I think I have tree shaked correctly.

Also, I'm experiencing that the website renders two times. I don't know if that could be connected to this, and to the slow rendering time of the server and client bundle, but it seems to render the first time statically from server, without rendering again in any way, like a "pre-render" which is not functional, then the page goes blank for a little, then re-renders components from client and they are fully functional, all in like 3 seconds.

Many thanks!

P.S. oh, to all that may be asking, yeah, everything is in production mode, and Next uses getServerSideProps in all pages.

  [1]: https://i.stack.imgur.com/DQO5M.png

**Comment by Vadorequest** [comment]
You must be loading quite a lot of css/js on the frontend, maybe are you loading packages on the frontend that are only used by the backend?

Your main issue is the "shared with all" of ~600kB, it's quite a lot, and that's why all your pages are big. Check what you import in your _app.

As a comparison, [NRN](https://github.com/UnlyEd/next-right-now/pull/68/) includes a lot of stuff, but isn't that big. (each page is about 500kB still, but as I said it does a lot of stuff)

![image](https://user-images.githubusercontent.com/3807458/83802071-6fe2b100-a6aa-11ea-9d6c-2c7dd6a026bd.png)

**Comment by DanielVip3** [comment]
Mh, that's true, but the fact is that my app uses, in his _app.js, both Navbar and Footer components which call 4 different libraries and some Material-UI components which I use in the entire website. But it should not be THAT heavy.
Tomorrow I'll post all the imports made from _app.js recursively in all imported files.

**Reply by Vadorequest** [reply]
_You should reply in the thread, easier to follow._

https://bundlephobia.com/result?p=material-ui@0.20.2 MUI is ~100kB.

**Reply by DanielVip3** [reply]
Yeah, thanks!
Anyway, MUI is ~100 KB, but that shouldn't be the issue itself, so I'm just gonna look for weird imported packages right and then. I'll let you know in the next 2-3 hours.

**Reply by DanielVip3** [reply]
I tried importing dynamically everything. Apart from the fact that now the page loads progressively, and that sucks, anyway;
I noticed that Google Lighthouse says that lots of unused JS is called, including Recaptcha:
https://i.imgur.com/RolIXNj.png
I don't use Repcatcha in any page except the signup one. So, that looks like, on the homepage, some of the signup page is imported? I can't understand. Is that the normal behaviour?

**Comment by wereHamster** [comment]
Do you have `"sideEffects": false` in your package.json?

**Reply by DanielVip3** [reply]
No, I don't- actually my package.json only contains dependencies, devDependencies, proxy, private, version, name and scripts.

**Reply by lauchness** [reply]
This was my issue, thanks!

**Reply by bryanprimus** [reply]
this reduced my bundle size significantly. but I'm not sure why

**Reply by IonelLupu** [reply]
"sideEffects": false doesn't work in my case: https://github.com/vercel/next.js/discussions/60206
